### ISSUE TITLE
wrong scroll with mouse scroll buttons and background tabs on

### ISSUE THREAD

**Issue by tex** [The-Compiler/qutebrowser#395]
When background tabs are true, middle click on a link causes the target to be opened on background tab, then scrolling with the current page by mouse scroll is incorrect - qutebrowser first moves to begining of the bage, then scroll movement is applied.
It looks like when scrolling with a mouse the position of the background tab is applied on the current tab. Scrolling with keys is ok though.

Anyway - great project!

**Comment by The-Compiler** [comment]
@tex how are things looking with Qt 5.5?

**Comment by tex** [comment]
Well, recently I'm getting this error on start and it quits:
ERROR: Invalid value 'http://www.malwaredomainlist.com/hostslist/hosts.txt,http://someonewhocares.org/hosts/hosts,http://winhelp2002.mvps.org/hosts.zip,http://malwaredomains.lehigh.edu/files/justdomains.zip,http://pgl.yoyo.org/adservers/serverlist.php?hostformat=hosts&mimetype=plaintext,' - values may not be empty!
Traceback (most recent call last):
  File "/usr/lib/python3.4/site-packages/qutebrowser/config/config.py", line 125, in init
    config_obj = ConfigManager(confdir, 'qutebrowser.conf', app)
  File "/usr/lib/python3.4/site-packages/qutebrowser/config/config.py", line 234, in __init__
    self._validate_all()
  File "/usr/lib/python3.4/site-packages/qutebrowser/config/config.py", line 371, in _validate_all
    opt.typ.validate(interpolated)
  File "/usr/lib/python3.4/site-packages/qutebrowser/config/configtypes.py", line 1181, in validate
    raise configexc.ValidationError(value, "values may not be "
qutebrowser.config.configexc.ValidationError: Invalid value 'http://www.malwaredomainlist.com/hostslist/hosts.txt,http://someonewhocares.org/hosts/hosts,http://winhelp2002.mvps.org/hosts.zip,http://malwaredomains.lehigh.edu/files/justdomains.zip,http://pgl.yoyo.org/adservers/serverlist.php?hostformat=hosts&mimetype=plaintext,' - values may not be empty!

So I stopped using qutebrowser. Will tests it again if this gets fixed, recently I'm very busy in my work, family and I don't have too much time to play with things as I used to have...

**Comment by The-Compiler** [comment]
That's not a bug - you have a trailing comma in your setting value for `content -> host-block-lists` which shouldn't be there.

**Comment by tex** [comment]
The thing is that I'm unable to find out where exactly is the config file. I can see a name of the config file in the log but not path. My ~/.config/qutebrowser is empty...

**Comment by tex** [comment]
Yep, it tries to read qutebrowser.conf form ~/.config/qutebrowser/ which is empty and this is result. Not very helpful message to user...

**Comment by tex** [comment]
With ```qutebrowser -c ""``` it actually runs. You should provide a better user experience to new users...

**Comment by tex** [comment]
I had an old version. Arch's qutebrowser-git package somehow failed to work for me.
No problem with starting now, but the #395 issue still the there. Less frequently but still there. It happend only once (scrolled to bottom a large page by mouse, then middle-click on a link, it opened in backround, scrolled up a little and it scrolled the page right to top.

**Comment by tex** [comment]
Closing now, if I find how to reproduce it reliably I open a new issue.

**Comment by zuglufttier** [comment]
I have that problem, too. Is there something I can do to debug this? I use a Thinkpad X1 Carbon 2015.

The problem only arrises using the toupad but it does not happen, when I connect a real mouse via USB or I'm using the trackpoint. This seems to be related how the events from the different input methods are interpreted.

**Comment by The-Compiler** [comment]
@zuglufttier Hmm, odd - some questions:

* What Qt and qutebrowser version? (`:open qute:version`)
* What OS/distribution?
* Can you reproduce it constantly?
* Are you sure it's the same issue, i.e. scrolling issues *on the page you're on*, not the new page?
* Can you also reproduce it by middle-clicking somewhere else on the page (i.e. not a link)?
* Can you maybe do a screencast using byzanz or recordmydesktop so I can see what's going on precisely?

**Comment by zuglufttier** [comment]
* qute:version output:

> qutebrowser v0.3.0
> Git commit: v0.3.0-634-g3a5cd8f (2015-09-03 06:51:15 +0200)
> 
> CPython: 3.4.3
> Qt: 5.5.0, runtime: 5.5.0
> PyQt: 5.5
> Style: QGtkStyle
> Desktop: None
> sip: 4.16.9
> colorlog: no
> colorama: no
> pypeg2: 2.15
> jinja2: 2.8
> pygments: 2.0.2
> yaml: 3.11
> Webkit: 538.1
> Harfbuzz: old
> SSL: OpenSSL 1.0.2d 9 Jul 2015
> 
> Frozen: False
> Platform: Linux-4.1.6-1-ARCH-x86_64-with-arch-Arch-Linux, 64bit
> OS Version: arch, Arch, Linux
> 
> --- /etc/os-release ---
> NAME="Arch Linux"
> ID=arch
> PRETTY_NAME="Arch Linux"
> ANSI_COLOR="0;36"
> HOME_URL="https://www.archlinux.org/"
> SUPPORT_URL="https://bbs.archlinux.org/"
> BUG_REPORT_URL="https://bugs.archlinux.org/"
> 
> 
> 
> --- /etc/lsb-release ---
> LSB_VERSION=1.4
> DISTRIB_ID=Arch
> DISTRIB_RELEASE=rolling
> DISTRIB_DESCRIPTION="Arch Linux"
> 
> 
> --- /etc/arch-release ---
> Arch Linux release

* As you can see, it's a regular arch linux, recently updated
* Yes, I can reproduce it constantly
* Yes, it happens on the page I stay on, not in the background tab
* Yes, only middle clicking will result in the same behaviour
* Maybe later ;-)

**Comment by tex** [comment]
That's me again. Started to use qutebrowser again as I am disappointed with Firefox, but this error is there and I can reproduce it constantly.
When I open tab with middle button, switch to the opened tab, view the page, then back to original page, there I can scroll with keys with no issue. But when I scroll on touchpad it instantly moves page at the top.

**Comment by The-Compiler** [comment]
@tex What machine are you using? Is it by any chance also a Thinkpad X1 Carbon like @zuglufttier has?

Unfortunately it'll be hard to debug this without being able to reproduce it myself... Is anyone of you either close to the Swiss border or going to the 32c3 by any chance?

**Comment by tex** [comment]
Sorry, I have Dell E6430. :-) I'm in Czech Republic, not so far from Swiss, unfortunatelly no chance to visit it any soon.

**Comment by zuglufttier** [comment]
I live pretty far up north in Germany, so it's rather unlikely ;-) I recently changed my desktop environment from spectrwm to a full fledged gnome installation but that did not change anything.

I finally came around to record this behaviour: https://drive.google.com/file/d/0B6riaaKSZizoMkdOeXZrRDdXdmc/view?usp=sharing

As you can see, it only happens when I use the touchpad and not when I'm using hints to open the links in background tabs. Using the trackpoint or a usb mouse will not show this problem!

**Comment by The-Compiler** [comment]
Thanks, that definitely helps with seeing what's going on.

I'm in Hamburg for the 32c3 on 26th-31st December - any chance we could meet then, in case Hamburg is somewhere close to you? I have the feeling that this is very easy to fix once I can actually reproduce it and test some solutions, and it'd definitely easier than debugging this remotely...

I'll likely meet someone with an X1 carbon tomorrow, so maybe I can try reproducing it there.

Do you have any special settings (like via `xinput`) for the trackpoint? I use that to enable middle-click scrolling, I'll try without that later too.

**Comment by zuglufttier** [comment]
I'm not that far from Hamburg actually but I might not be in Germany at all at that time ;-)

I had those settings to enable middle-click scroll but I also tried without. I recently did a complete reinstallation and the behaviour is the same.

I also think it should be easy to fix... I used `xev` to get the events that happen in the background. This happens when I use the trackpad:

>ButtonPress event, serial 37, synthetic NO, window 0x1c00001,
>    root 0xf6, subw 0x0, time 4625283, (251,343), root:(1720,539),
>    state 0x0, button 4, same_screen YES
>
>ButtonRelease event, serial 37, synthetic NO, window 0x1c00001,
>    root 0xf6, subw 0x0, time 4625283, (251,343), root:(1720,539),
>    state 0x800, button 4, same_screen YES

And this is using the trackpoint or a usb mouse:

>ButtonPress event, serial 37, synthetic NO, window 0x1c00001,
>    root 0xf6, subw 0x0, time 4761015, (116,109), root:(116,148),
>    state 0x0, button 5, same_screen YES
>
>ButtonRelease event, serial 37, synthetic NO, window 0x1c00001,
>    root 0xf6, subw 0x0, time 4761015, (116,109), root:(116,148),
>    state 0x1000, button 5, same_screen YES

One time it's button 4, the other time button 5.

**Comment by The-Compiler** [comment]
Oooh, *touchpad*! I somehow missed that for some reason... I have mine deactivated, but I can try with that later.

How do you do the middle-click exactly then? I only have two (pseudo-)buttons for the touchpad, and three for the trackpoint: 

![touchpad](https://cloud.githubusercontent.com/assets/625793/9872523/2449e200-5b9a-11e5-9bab-aeb24ddb2072.jpg)

**Comment by zuglufttier** [comment]
I just use that middle mouse button from the trackpoint. The events i captured with `xev` are from the scrolling event.

**Comment by The-Compiler** [comment]
Heh - even with it re-enabled in the BIOS I can't get my touchpad to work...

I'll see if I can borrow someone's machine to test tomorrow when I'm in the local hackerspace.

**Comment by zuglufttier** [comment]
Maybe it's just something like `synclient TouchpadOff=0`?

**Comment by The-Compiler** [comment]
This seems to be a Qt thing, I can even reproduce it in `qpdfview`...

I added a workaround which just ignores the first scroll event after a click - please let me know if that helped!

**Comment by tex** [comment]
It is gone! Thanks!

**Comment by zuglufttier** [comment]
I can confirm that, too! Very nice. Not so nice that it's happening in Qt ;)

Thanks anyway!

**Comment by The-Compiler** [comment]
@tex @zuglufttier I just pushed some changes moving this code around - if you can verify that the workaround still works in the latest master that'd be most appreciated, as I don't have the hardware to test it around right now :wink:

**Comment by tex** [comment]
@The-Compiler It seems to work correctly.

**Comment by zuglufttier** [comment]
A quick test showed no problems.

**Comment by haasn** [comment]
```irc
13:46 <haasn> The-Compiler: I figured out how to reproduce my “qutebrowser sometimes decides to scroll up when I use the mouse wheel” bug
13:46 <haasn> it always happens after I draw a mouse gesture
13:47 <haasn> (which I do with `easystroke`)
13:47 <haasn> so it's like draw invalid gesture -> scroll down
13:47 <haasn> and it jumps up a few lines on the first scroll event
13:47 <haasn> doesn't happen with the keyboard
13:47 <haasn> no idea what causes it
13:47 <haasn> but at least now I know how to reproduce it
14:19 <The-Compiler> haasn: FWIW I use easystroke too
14:25 <haasn> The-Compiler: how do you have easystroke set up? this is what I'm doing https://0x0.st/2-n.png
14:26 <The-Compiler> haasn: and you have your mousewheel button set up as easystroke button?
14:26 <haasn> and if I open an arbitrary web page with --temp-basedir and do <press right click> <squiggle around a bunch> <release right click> <scroll down once> it jumps to the top of the page
14:26 <haasn> The-Compiler: yes, but it only counts as a gesture when I have RMB held down
14:27 <haasn> which is not the case by the time I emit the scroll event
14:34 <The-Compiler> https://codereview.qt-project.org/#/c/175683/ - nice, no 2 days after I requested it :)
14:34 <The-Compiler> haasn: hmm, wait, that sounds similar
14:34 <The-Compiler> err, familiar
14:36 <The-Compiler> haasn: https://github.com/The-Compiler/qutebrowser/issues/395 https://github.com/The-Compiler/qutebrowser/commit/8ce7f76b1603a02bca656e6ad06efb25115b3169
14:42 <haasn> The-Compiler: that does sound incredibly similar
14:42 <The-Compiler> haasn: mind adding a comment to the bug with all the infos you have? Then I'll reopen it and look at it again whenever I have a chance
```

**Comment by The-Compiler** [comment]
I've seen this happen again recently with QtWebEngine myself...
